include/boost/url/impl/url.ipp

```diff
@@ -2042,26 +2044,112 @@ resolve(
 //
 //------------------------------------------------
 
-url&
+void
 url::
-normalize()
+to_lower_impl(int id) noexcept
 {
-#if 0
-    // scheme
+    char* it = s_ + offset(id);
+    char const* const end = s_ + offset(id + 1);
+    if(it < end)
     {
-        char* it = s_
-            + offset(id_scheme);
-        auto last = it + offset(id_user);
-        if(it < last)
+        while(it < end)
         {
-            --last;
-            while(it != last)
+            if (*it != '%')
             {
-                // to upper
+                *it = grammar::ascii_tolower(
+                    *it);
+                ++it;
+                continue;
             }
+            it += 3;
         }
     }
-#endif
+}
+
+url&
+url::
+normalize_scheme()
+{
+    to_lower_impl(id_scheme);
+    return *this;
+}
+
+url&
+url::
+normalize_authority()
+{
+    // normalize host
+    if (host_type() == host_type::name)
+    {
+        static constexpr auto reg_name_chars =
+            unreserved_chars
+            + '-' + '.';
+        normalize_octets_impl(id_host, reg_name_chars);
+        to_lower_impl(id_host);
+    }
+
+    // normalize password
+    static constexpr auto pass_cs =
```

Comment:
These charsets are coming from other `set_encoded_*` functions or from the corresponding grammar rule. It's probably appropriate to define these in a single place but I don't know where the best place for that would be. 

I also noticed Javascript's URL class 1) normalizes all URLs, and 2) is more permissive for `reg_name_chars` (it accepts subdelim_chars). For instance, the host would decode a `!`:

```js
new URL('http://%68examPle%21.com:80/path/../index.html?gliu%5B%5D=1');
```

```js
URL {
  href: 'http://hexample!.com/index.html?gliu%5B%5D=1',
  origin: 'http://hexample!.com',
  protocol: 'http:',
  username: '',
  password: '',
  host: 'hexample!.com',
  hostname: 'hexample!.com',
  port: '',
  pathname: '/index.html',
  search: '?gliu%5B%5D=1',
  searchParams: URLSearchParams { 'gliu[]' => '1' },
  hash: ''
}
```

Reply:
if you want to share the charsets put it in `<boost/url/detail/charsets.hpp>` and make sure it is not visible publicly (i.e. only included from .ipp files)